### ISSUE TITLE
Run tests under TravisCI

### ISSUE THREAD

**Issue by mjs** [TheCacophonyProject/cacophony-api#71]

**Comment by ben-biddington** [comment]
Hey can you confirm all tests are passing locally?

**Comment by ben-biddington** [comment]
I am trying to stub out `openS3` so I am sure I have made a mistake somewhere, but just in case

**Comment by clare** [comment]
Hi Ben, 

Welcome to Cacophony - great to have you helping.  

I'll get the latest and double check for you, but I think they are working right now.   

FYI the tests are not unit tests - they are integration level tests and we normally run them against an instance of server running against Minio (openS3) and a postgres database on our machines.   The instructions are here https://github.com/TheCacophonyProject/cacophony-api/blob/master/test/README.md.  

Have you remembered to set up the super user?   This isn't automated yet and would be a great place to start.   

Otherwise can you give me a better idea of the errors you are getting and I might be able to help more.

FYI A few of the tests do test saving to S3 and getting the files back so that could cause a few issues if it wasn't working but I wouldn't expect all the tests to fail then. 

Cheers
Clare

**Comment by clare** [comment]
Yes, just confirmed.   All tests are passing right now.

**Comment by ben-biddington** [comment]
Draft [here](https://github.com/ben-biddington/cacophony-api/blob/docker/models/util/util.js#L322)

**Comment by mjs** [comment]
Hi @ben-biddington. To be clear, we're not using S3 itself (for the integration tests or in production) but a S3 clone. We're using the excellent [Minio](https://minio.io/downloads.html#download-server-linux-x64) which is very easy to operate (single binary, one command to run).

It *should* be fairly easy to run Minio inside a TravisCI test run.

What we really need is to automate the steps required to set up for running tests both to make things easier on developer machines but also to facilitate running the tests automatically with TravisCI.

**Comment by mjs** [comment]
Eventually it'll also be nice to have a JavaScript based unit tests. Then, a fake storage might be useful.

**Comment by ben-biddington** [comment]
Hmm just seems like a lot of hassle for something not that interesting (file storage). It is also quite tightly stuck where an abstraction could be.

Will give it a go

**Comment by mjs** [comment]
OK, if you're keen then give it a go.

**Comment by mjs** [comment]
It looks like you've made great progress. Thank you! Having the tests running under TravisCI will be a big step forward. A Docker build is nice to have to.

It looks like you've exposed a bug which causes the API handler to return headers twice. I'll take a look at that. Ironically, I'm working on a big change to the way audio recordings work. This bug likely doesn't  even exist in that branch.

As to why the problem is happening for you when it doesn't for me, I'm not sure. Did you say that the test admin user is set as a super user?

**Comment by mjs** [comment]
I haven't been able to replicate the double response issue you're seeing but I think I've found the problem code anyway. This should fix it: https://github.com/TheCacophonyProject/cacophony-api/pull/99

This code is all about to go away (this week hopefully) but it's good to get it sorted anyway.

**Comment by ben-biddington** [comment]
@mjs There is that yeah, but isn't it due to the recording not being found?

Seems to upload okay but then not come back.

Is there something with the audio processing I am missing?

**Comment by mjs** [comment]
Could it be that the test admin user isn't a superuser so it can't see the just uploaded audio file?

**Comment by ben-biddington** [comment]
Could be yeah, will look into that, I am [seeding](https://github.com/ben-biddington/cacophony-api/blob/master/test/db-seed.sql) it, but there must be an error

**Comment by mjs** [comment]
The seed SQL looks OK to me.... 

Also, I wasn't able to replicate the error even when I intentionally left the admin user with super=false. There must be something environmental going on.

Can you run just one of the failing tests can provide the API server logs for just that test run?

**Comment by ben-biddington** [comment]
Yep, that output up there is for `TestUserAudio#test_can_delete_audio`

**Comment by ben-biddington** [comment]
The other thing is the `test-group` group -- what needs to happen with that?

**Comment by mjs** [comment]
You actually shouldn't need to add "test-group" in the seed. I'm pretty sure the tests do that themselves (and do that using the test admin user).

**Comment by ben-biddington** [comment]
I guess you could try the container and see if you can diagnose?

https://github.com/ben-biddington/cacophony-api/blob/master/docker-start.sh

**Comment by ben-biddington** [comment]
Starts server on 1080

**Comment by ben-biddington** [comment]
Very cute port number choice, by the way :rat:

**Comment by mjs** [comment]
Yes that was @clare's little easter egg :)

I'll try out the container locally.

**Comment by mjs** [comment]
I'm having fun getting the container to even start. The main issue turned out to be that the ubuntu version wasn't specified so it was using bionic pre-release for me which includes PostgreSQL 10 and breaks assumptions about PostgreSQL config files.

With that fixed it's now getting further but the server fails to start due to a missing bcrypt dependency. Sorting through it.

**Comment by ben-biddington** [comment]
Yeah lots of trial and error with this stuff

**Comment by ben-biddington** [comment]
Onto the final travis step now which is the python setup.

I have it working locally but can't remember how I installed it.

It fails like this: https://travis-ci.org/ben-biddington/cacophony-api/builds/391568650

I get same bcrypt error with ubuntu 16.04, I think trusty should work (14.04)

**Comment by mjs** [comment]
I almost have the issues fixed. Expect a PR for your branch soon.

Looks like TravisCI is failing because Python 2 is being used. The tests using Python 3.

**Comment by ben-biddington** [comment]
Okay: now have the same 4 failures: https://travis-ci.org/ben-biddington/cacophony-api/builds/391569228

**Comment by mjs** [comment]
Progress :)  

Hopefully I can figure out what's going on with those.

**Comment by mjs** [comment]
OK container working and the tests are failing in the same way for me.

**Comment by ben-biddington** [comment]
@mjs  Just push your dockerfile changes if you like, have added you as collab at https://github.com/ben-biddington/cacophony-api

**Comment by mjs** [comment]
I have the tests passing. For some reason including the test group in the seed appears to be the problem. Just confirming that's really the case.

**Comment by mjs** [comment]
Thanks for giving me access to your repo. I'll push soon.

**Comment by mjs** [comment]
Fixes pushed

**Comment by ben-biddington** [comment]
Down to one failure after removing seed: https://travis-ci.org/ben-biddington/cacophony-api/builds/391576091

**Comment by mjs** [comment]
Have you tried with my fixes to the image? All tests were passing for me.

Note that I switched it back to 16.04 as that's what we run in production (my fixes means that 16.04 works)

**Comment by ben-biddington** [comment]
Actually, sorry, it is passing: https://travis-ci.org/ben-biddington/cacophony-api/builds/391577065

**Comment by ben-biddington** [comment]
Yep passing locally too :smoking:

**Comment by ben-biddington** [comment]
Will give travis/docker another go

**Comment by mjs** [comment]
Awesome! Thank you!

I don't mind whether we use Docker with Travis or not. Might be nice but whatever's easiest really.

**Comment by ben-biddington** [comment]
Yeah it's a bit duplicative now -- docker would solve it

**Comment by ben-biddington** [comment]
Yep: docker is a goer: https://travis-ci.org/ben-biddington/cacophony-api/builds/391590117

Will submit the changes tomorrow :green_apple:
